python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py

```diff
@@ -105,12 +112,58 @@ class DygraphToStaticAst(gast.NodeTransformer):
     """
 
     def get_static_ast(self, root):
-        # save root for some analysis may need global AST 
+        # save root for some analysis may need global AST
         self.root = root
         self.static_analysis_root = StaticAnalysisVisitor(
             root).get_node_wrapper_root()
+        self.static_analysis_root = root
+        # record all created ast.functionDef in control flow statement
+        self.new_func_nodes = []
+        self.decorate_func_name = None
         self.transfer_from_node_type(self.static_analysis_root)
-        return self.static_analysis_root
+        return self.static_analysis_root, self.decorate_func_name
 
     def transfer_from_node_type(self, node):
-        print("Not implemented")
+        self.visit(node)
+        # add new ast.funcDef of `if/else`
+        node.body = self.new_func_nodes + node.body
+
+    def visit_If(self, node):
```

Comment:
Maybe I didn't talk to you clearly about the thinking in my mind.

I thought it is better to have something like this in `transfer_from_node_type`:

if "current_node is if node": # the case you are handling
   then do a visit at current_node

But in this PR, you are implementing visit_XXX on all cases. For example, no matter whether the AST has Call node, you replace `YYY.numpy` to `YYY`, this may be right in your `if` case, but may not be true in all cases (for example, a user-defined class which has a field named `numpy`).

There will be two advantage if you can visit only the `if` node:
1. The numpy case will be limited into your `if condition`
2. Other RD may also change visit_Call and you won't have conflict. 

Based on the reasons, my suggestion on current PR will be:
at visit_If, make a you-defined visitor/transformer, then do your visit in that `if` node only, not the whole AST.

Reply:
Great advice! 
I extract the code related to `if/else` transformation into `class IfElseTransformation`. 

There are two things to note：
1. Only `ast.If` will  be parsed and modified in `visit_if`， even though `root` be required.(We need root to parse parameters of nested `true_fn` and `false_fn`)

2. `tensor.numpy()[]` is an interface of Dygraph and may be considered as `Basic API`, I will remove it after yamei's is merged. It's needed for unittest.

Reply:
OK, since tensor.numpy() will be a Basic API, and my PR to analysis type is merged. Can we only remove the .numpy when the variable is tensor? But this will be Yamei's work :-)

Reply:
Sorry, considering the parameters and returns, the `if` node requires `root`, thank you! But let's split in a different transformer for conflict reason

Reply:
> OK, since tensor.numpy() will be a Basic API, and my PR to analysis type is merged. Can we only remove the .numpy when the variable is tensor? But this will be Yamei's work :-)

I will  merge this function into BasicApiTransformer in next PR.